ml-agents/mlagents/trainers/torch/utils.py

```diff
@@ -333,3 +338,93 @@ def soft_update(source: nn.Module, target: nn.Module, tau: float) -> None:
                     alpha=tau,
                     out=target_param.data,
                 )
+
+    @staticmethod
+    def create_residual_self_attention(
```

Comment:
I do not know how I feel about these two methods. I think they would look better as a module given how they are used.

Reply:
Maybe we can combine this with a change to NetworkBody to also satisfy the comment that MultiAgentNetworkBody should reuse NetworkBody. The right solution may be to rename NetworkBody to Encoders that has the functionality of these methods and move the linear layer to the Actor/Critic. Then the MultiAgentNetBody may be able to reuse the encoders.

Reply:
reference https://github.com/Unity-Technologies/ml-agents/pull/5005#discussion_r591813207